### PR TITLE
Add OmicsUri region, account_id, and default file_name support

### PR DESCRIPTION
**Pull request by ahenroid**
*Issue #, if available:* N/A

*Description of changes:*
- Parse region and account ID segments of the URI and add corresponding attributes to OmicsUri
- Allow URIs with no file_name segment, adding in the default file, SOURCE1 for a ReadSet or SOURCE for a Reference
- Tighten URI parsing by using a regular expression
- Extend and add unit tests to bring URI parser coverage to 100%

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: omics/uriparse/uri_parse.py [hunk 1]
```diff
@@ -30,8 +30,15 @@ def __init__(self, omics_uri):
         if url_parts.scheme != self.SCHEME:
             raise ValueError(f"Invalid URI scheme, expected {self.SCHEME}")
 
-        if self.NETLOC not in url_parts.netloc:
-            raise ValueError(f"Invalid URI netloc, expected {self.NETLOC}")
+        netloc_parts = url_parts.netloc.split(".")
+        if len(netloc_parts) != 5 or netloc_parts[1] != self.NETLOC:
```
Comment (rforsyth) [inline omics/uriparse/uri_parse.py:34]: Have you considered using Regex Capture Groups to validate and parse the URL?  Seems like that could handle edge cases like unusual characters in the Account ID position and be more sophisticated than `split` and `strip`.

Not a requirement to merge - just worth discussing.

**Review by rforsyth — DISMISSED** [review]
This looks like a good enhancement in general; just had a question about whether it makes sense to do some refactoring with Regex.

File: omics/uriparse/uri_parse.py (hunk 1 above)
Comment (ahenroid) [reply omics/uriparse/uri_parse.py:34]: Yes, I was also worried about the open-ended parsing we're doing here and would prefer a regex. I'll update with those changes.

**Review by tneely — APPROVED** [review]

**Comment by ahenroid** [comment]
I've updated this PR with the reviewer recommendation to validate & parse URIs with a regular expression

File: omics/uriparse/uri_parse.py [hunk 2]
```diff
@@ -14,45 +14,30 @@ class OmicsUri:
     ex: omics://429915189008.storage.us-east-1.amazonaws.com/1981413158/readSet/5346184667/source1
     ex: omics://429915189008.storage.us-east-1.amazonaws.com/1981413158/reference/5346184667/source
     """
-
-    SCHEME = "omics"
-    NETLOC = "STORAGE"
-    DEFAULT_FILE_NAME = None
+    URI_REGEX = r"omics://(\d{10,12})\.storage\.([a-z]{2}-[a-z-]{4,}-\d+)\.amazonaws\.com/(\d{10,36})/(readSet|reference)/(\d{10,36})(/(source[12]?|index))?$"
```
Comment (rforsyth) [inline omics/uriparse/uri_parse.py:17]: According to a quick check on the regex - looks like the slashes should be escaped?

File: omics/uriparse/uri_parse.py [hunk 3]
```diff
@@ -14,45 +14,30 @@ class OmicsUri:
...
-            raise ValueError(
-                f"Invalid URI path for reference, {self.resource_type} is not a valid URI type"
-            )
-
-        if (
-            self.file_name
-            not in OMICS_URI_TYPE_FILENAME_MAP[OmicsFileType(self.resource_type)].list()
-        ):
-            raise TypeError("URI file is unsupported.")
+        uri_match = re.match(self.URI_REGEX, omics_uri, re.IGNORECASE)
```
Comment (rforsyth) [inline omics/uriparse/uri_parse.py:25]: So much better than splitting the string!  :)  Thanks for making the switch to regex.

**Review by rforsyth — APPROVED** [review]

**Pull request merged**